### PR TITLE
feat(template-compiler): lwc external directive

### PR DESCRIPTION
**Pull request by ekashida**
## Details

https://rfcs.lwc.dev/rfcs/lwc/0118-native-web-components-support

Initial implementation which only sets attributes.

## Does this pull request introduce a breaking change?

No

## Does this pull request introduce an observable change?

No

## GUS work item

W-11610725

### CODE REVIEW

File: packages/@lwc/template-compiler/src/shared/types.ts [hunk 1]
```diff
@@ -170,6 +170,10 @@ export interface Element extends AbstractBaseElement {
     type: 'Element';
 }
 
+export interface External extends AbstractBaseElement {
```
Comment (rui-rayqiu) [inline packages/@lwc/template-compiler/src/shared/types.ts:173]: Just for my understanding, what's the difference between having this `External` as BaseElement vs. using `ExternalDirective extends Directive<'External'>`?

Comment (ekashida) [reply packages/@lwc/template-compiler/src/shared/types.ts:173]: Maybe we should rename it but `External` is the type for the element with the directive and `ExternalDirective` is the type of the directive itself (i.e., `lwc:external`).

Comment (rui-rayqiu) [reply packages/@lwc/template-compiler/src/shared/types.ts:173]: Ok, but we don't really need `ExternalDirective` as one of `ElementDirective` type because it is not used anywhere and only `ElementDirectiveName.External` would be enough?

File: packages/@lwc/template-compiler/src/parser/attribute.ts [hunk 2]
```diff
@@ -204,6 +204,12 @@ export function isAttribute(element: BaseElement, attrName: string): boolean {
         );
     }
 
+    // External custom elements default to setting data as attributes. These might be set as
+    // properties during runtime, depending on runtime heuristics.
```
Comment (rui-rayqiu) [inline packages/@lwc/template-compiler/src/parser/attribute.ts:208]: I guess the runtime part will be in a separate PR?

**Review by rui-rayqiu — APPROVED** [review]

**Review by caridy — COMMENTED** [review]
Looking good, waiting for the attribute/prop logic though.

File: packages/@lwc/template-compiler/src/parser/attribute.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ekashida) [reply packages/@lwc/template-compiler/src/parser/attribute.ts:208]: No, for each attribute, we would check if a corresponding property exists and if so, set that instead. Unfortunately we know nothing during compile time so I defaulted to attributes since we will only be setting attributes for SSR.

File: packages/@lwc/template-compiler/src/shared/types.ts (hunk 1 above)
Comment (ekashida) [reply packages/@lwc/template-compiler/src/shared/types.ts:173]: @rui-rayqiu I initially defined the `ExternalDirective` type because I was interacting with it by adding it to the AST, but after I introduced the `ExternalComponent` type instead, I ended up not including the directive in the AST.

**Comment by ekashida** [comment]
@caridy attr/prop is coming next
